### PR TITLE
Use the build cache in CircleCI

### PR DESCRIPTION
**Pull request by yschimke**
TL;DR it probably saves a minute per build on average, and we could debug to improve.

Aggressive build caching within gradle and circleci.  Not sure we want this, it makes re-running a flaky test potentially awkward. It should be safe, but ultimately we want any snapshot or tagged release publishing step to do a fresh build I assume.

Normal: 5:40
https://circleci.com/workflow-run/7c6cd465-9609-41bd-b63c-d98334c9a027
image

Cache First Run: 5:41
https://circleci.com/workflow-run/a69206f5-ef50-415b-ba90-b0f97fefa33b
image

Second Run: 4:53
https://circleci.com/workflow-run/0a782b8a-5db8-4f4b-b76d-9792d9fef0fd
image

### CODE REVIEW

**Comment by yschimke** [comment]
Sorry, should be a draft PR

**Comment by yschimke** [comment]
n.b. times should get better as a master build should populate the cache for branch builds.  We can have a cache key command that generates a cache key from all build.gradle files, maybe combined with the current date.

Otherwise the cache becomes less relevant over time

image

**Comment by JakeWharton** [comment]
There shouldn't really be a need to use the date. I mean we shouldn't even
need to use the build file hash either except the cache would just grow
indefinitely. Maybe just the year and month as a key?

**Comment by yschimke** [comment]
https://guides.gradle.org/using-build-cache/

Lot of things to investigate e.g. debug the internal cache keys, since the environment variables probably kill the test result caching

image

**Comment by yschimke** [comment]
@JakeWharton what would you be comfortable caching across PRs?  e.g. would you like it if gradle only ran the tests for modules we had changed?  Should the target be the 10-20 second build & test times you can often experience locally?

**Comment by yschimke** [comment]
Also, once you have written the cache with a key it is immutable.  And restores are prefix based, so it's nuanced in an awkward way.

**Comment by JakeWharton** [comment]
Compilation for changed modules only. Tests for changed or downstream
(i.e., depends on a re-compiled module, potentially transitively).

**Comment by yschimke** [comment]
OK, leave open and I'll dig deeper tomorrow.

**Comment by yschimke** [comment]
With a change in a smaller test module, it's down to 2 minutes for a followup build

https://circleci.com/workflow-run/2f31ac40-681f-4d5f-bd32-d688cf512232

image

image
